### ISSUE TITLE
Add support for Notification Actions (added in Chrome 48)

### ISSUE THREAD

**Issue by bengotow** [electron/electron#4424]
It would be really great if Electron supported the "notification actions" feature added in Chrome 48:

https://www.chromestatus.com/features/5906566364528640
https://developers.google.com/web/updates/2016/01/notification-actions?hl=en

Essentially, you say:

```
new Notification("123", {title: "123", silent: true, actions: [{action: 'A', title: 'A'}, {action: 'B', title: 'B'}]})
```

To show the notification, and then listen for 'notificationclick' DOM event, which gives you the name of the chosen action.

**Comment by bengotow** [comment]
It's worth noting that the Chromium implementation uses their own notification UX and not the system notification center. Mac OS X has the equivalent of notification actions (up to two actions), so the existing API could be mapped onto Mac OS X. I'm not sure about other platforms!

![image](https://cloud.githubusercontent.com/assets/1037212/12905797/187c9176-ce8f-11e5-8b69-dcb223f1b177.png)

**Comment by paulcbetts** [comment]
@bengotow:

Windows 10: Yes
Windows 8: No
Windows 7: lol
Unity (i.e. Ubuntu): No
non-Unity (i.e. Fedora): Yes

**Comment by felixrieseberg** [comment]
I'm also super into this, but I'd say that the implementation on both OS X and Windows is not trivial. For starters, the Windows 10 UWP Notification API is super rich and can do all kinds of fancy things, but it doesn't use callbacks, but instead activation arguments (which is cool, because one can launch other apps via notification, but the model makes it hard to implement in Electron). Good news: [Code is available](http://blogs.msdn.com/b/tiles_and_toasts/archive/2015/10/16/quickstart-handling-toast-activations-from-win32-apps-in-windows-10.aspx).

On OS X, things are a little bit less impossible, but still annoying (the model is super-limiting, as @bengotow already pointed out). However, if we conform to the limits, a callback is fairly simple: We already ask the `userNotificationCenter` for the activation; and the notification object has a `activationType` property. The whole thing is a bit silly, but simple enough - it's basically 'that one action clicked' or 'that other action clicked' ;-)

Basically, if somebody wants to have a coffee/beer session with me (Nylas? Slack? GitHub?), we could probably hack this out in three hours or so.

**Comment by bengotow** [comment]
Hey! I've started playing with this, and it looks like it may be more of a mess than I imagined. I was able to add a libchromiumcontent patch to remove this assertion which blocked calls to `new Notification...` with an actions array (https://chromium.googlesource.com/chromium/src/+/master/third_party/WebKit/Source/modules/notifications/Notification.cpp#82) and extend brightray to display Mac OS X notifications with the action titles:

![image](https://cloud.githubusercontent.com/assets/1037212/14071769/887055ea-f46d-11e5-99f1-d1e42c0cedac.png)

It looks like we may be blocked on the event-handling side, though. Chromium calls `PlatformNotificationService.DisplayNotification` with an instance of [content::DesktopNotificationDelegate](https://chromium.googlesource.com/chromium/src/+/2ee1f3c08f9d1e46bdeb42dc8435dd30d81f0ae6/content/public/browser/desktop_notification_delegate.h), not the (way more robust) [message_center::NotificationDelegate](https://chromium.googlesource.com/chromium/src.git/+/lkcr/ui/message_center/notification_delegate.h). It looks like the notifications API is unified on the JavaScript side but platform notifications and the "message center" are disconnected concepts.

Any ideas? 

One other caveat I discovered:

On Mac OS X, notifications with actions are "Alerts", which are sticky. To show "Alert"-style notifications, the app needs to be code-signed and include an item in it's Info.plist. Beyond being a bit of a pain, this means that effectively re-producing native behaviors might require additional APIs for programmatically dismissing notifications, which I don't think exist.

**Comment by bengotow** [comment]
This may be relevant: https://bugs.chromium.org/p/chromium/issues/detail?id=326539&can=2&start=0&num=100&q=&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified&groupby=&sort=

**Comment by Aluxian** [comment]
@bengotow If it helps, this is how I currently display native notifications on OS X (using nodobjc): https://github.com/Aluxian/Whatsie/blob/master/src/scripts/browser/components/native-notifier/impl-darwin.js

**Comment by koush** [comment]
Ah, I'd like this as well!

I attempted to get the serviceworker APIs working on electron, but notifications from a service worker do not show up.

Alternatively, chrome APP platform implements chrome.notifications.create, which may be worth looking at. Not sure if that is implemented differently from the serviceworker.newNotification.

**Comment by davies147** [comment]
"I attempted to get the serviceworker APIs working on electron, but notifications from a service worker do not show up."

+1

Even if it could fall-back to the standard Notification system, it would be helpful?

**Comment by codebytere** [comment]
Closing with merge of @sethlu's above PR.
